Added the framework for the OCS rules and alerts

anmolsachan opened the pull request:
This PR adds a framework for generating OCS rules and alerts. 

This is a follow-up PR on #733, #737 and #739. This PR should clear up how OCS rules yamls are created and how they will be maintained in future.

Doc to install dependencies and use : metrics/mixin/README.md

Signed-off-by: Anmol Sachan <anmol13694@gmail.com>

anmolsachan commented:
@umangachapagain @obnoxxx @jarrpa Please review.

anmolsachan commented:
> I have literally no idea how any of this works. Please include documentation (say under `metrics/docs`) to explain what all this means and how developers are supposed to use this. Links to external tutorials are also valid but probably not sufficient.

@jarrpa I understand and thanks for pointing that out. Since it is related to only "MIXIN' I have added the readme at metrics/mixin/README.md

jarrpa commented:
> > I have literally no idea how any of this works. Please include documentation (say under `metrics/docs`) to explain what all this means and how developers are supposed to use this. Links to external tutorials are also valid but probably not sufficient.
> 
> @jarrpa I understand and thanks for pointing that out. Since it is related to only "MIXIN' I have added the readme at metrics/mixin/README.md

That still doesn't tell me what any of this means or what it does. Also, it doesn't mention how to actually make changes, just hwo to generate artifacts after the actual work has been done.

anmolsachan commented:
> > > I have literally no idea how any of this works. Please include documentation (say under `metrics/docs`) to explain what all this means and how developers are supposed to use this. Links to external tutorials are also valid but probably not sufficient.
> > 
> > 
> > @jarrpa I understand and thanks for pointing that out. Since it is related to only "MIXIN' I have added the readme at metrics/mixin/README.md
> 
> That still doesn't tell me what any of this means or what it does. Also, it doesn't mention how to actually make changes, just hwo to generate artifacts after the actual work has been done.

@jarrpa I have added more info in the README file. Should make things more clear.

jarrpa commented on metrics/mixin/README.md:
```diff
@@ -0,0 +1,76 @@
+
+### To add new alerts/rules:
+
+* Add new alerts in 'metrics/mixin/alerts' directory in libsonnet format. Learn about Jsonnet and Libsonnet from https://jsonnet.org/.
```
This is not sufficient to even explain your example. For instance, I have no idea what `+::` does, as it is mentioned nowhere in the Jsonnet website. You can't assume that anyone coming to this repo has any idea how to work with jsonnet, or even prometheus mixins. The design doc is a bit too high-level and verbose to tell a developer how to modify this structure.

We're also missing and explanation of what the various files in this directory are for. For instance, why are there both `services-external.libsonnet` and `services.libsonnet` when they appear to be identical?

anmolsachan replied:
@jarrpa Its a new language and its not just us using it, It's a standard in Kubernetes monitoring world. Eg https://github.com/openshift/cluster-monitoring-operator/, https://github.com/kubernetes-monitoring/kubernetes-mixin. Just like I had to learn Golang to contribute to the OCS-operator, Jsonnet is a requirement to learn here. Also in OCS operator we have two active contributors for this, me and @umangachapagain and will be mostly. If more contributors have to contribute, they will have to learn.

Regarding the documentation, it's not as simple to get it, one has to do the whole tutorial to learn. Me and @umangachapagain have been through the pain. 

Regarding keeping two different files, I will give you an example: https://github.com/rook/rook/tree/master/cluster/examples/kubernetes/ceph/monitoring. If you have a look at the files in this repo you will see that external mode files have way lesser alerts than internal as they are no applicable there. 
We will have the same situation in OCS-operator as well, as the list of the alerts grows. Also, the OCS-operator alerts are not identical, there is a slight difference in the message and description fields which is because of the difference between the implementation of the internal and external mode of cephobjectstore.

jarrpa replied:
This is worlds different than just knowing Golang. But even there, I have a todo item to improve the on-boarding documentation so people can get started with minimal working knowledge in Go. Best I can tell, Jsonnet doesn't have nearlyt he same market penetration as Go either, so you can't just assume people will be able to pick it up with the same ease. I don't want to skip that here, especially because *I* need to know what's going on before I can approve anything.

If you have the relevant Jsonnet tutorials, link those in the README. Any maintainer who is neither you or Umanga should be able to look at this PR and understand what is happening. Links to the projects you just referenced probably wouldn't hurt, either.

umangachapagain commented on Makefile:
```diff
@@ -32,6 +32,7 @@ $(call add-crd-gen,ocsv1,./pkg/apis/ocs/v1,./deploy/crds,./deploy/crds)
 	gen-release-csv \
 	gen-latest-csv \
 	gen-latest-deploy-yaml \
+	gen-latest-prometheus-rules-yamls \
```
```suggestion
	gen-latest-alerts \
```

Just so everyone can understand

anmolsachan replied:
At the end of the days, its not just alerts, it will contain rules as well. By definition, they are together called Promethues Rules, that's why prometheus-rules-yamls, as rule yamls are the ones deployed.

umangachapagain commented on metrics/mixin/Makefile:
```diff
@@ -0,0 +1,29 @@
+# This Makefile should be used by developers to test generation of new new rules and alerts
```
Let's link it to the main Makefile so we can run these from root.

anmolsachan replied:
ack

umangachapagain commented on Makefile:
```diff
@@ -89,6 +90,10 @@ gen-latest-deploy-yaml:
 	@echo "Generating latest deployment yaml file"
 	hack/gen-deployment-yaml.sh
 
+gen-latest-prometheus-rules-yamls:
```
add lint checks to ensure we generate valid rules.

anmolsachan replied:
ack

umangachapagain commented on metrics/mixin/README.md:
```diff
@@ -0,0 +1,85 @@
+
+A set of Prometheus Rules & Alerts for OCS.
+
+The scope of this directory is to provide OCS specific Prometheus rule files using Prometheus Mixins.
```
```suggestion
The scope of this directory is to provide OCS specific Prometheus rule files using Monitoring Mixin.
```

anmolsachan replied:
ack

umangachapagain commented on metrics/mixin/alerts/services-external.libsonnet:
```diff
@@ -0,0 +1,27 @@
+  prometheusAlerts+:: {
+    groups+: [
+      {
+        name: 'cluster-services-alert.rules',
```
```suggestion
        name: 'external-cluster-services-alert.rules',
```

anmolsachan replied:
ack

umangachapagain commented on metrics/mixin/alerts/services-external.libsonnet:
```diff
@@ -0,0 +1,27 @@
+            },
+            annotations: {
+              message: 'Cluster Object Store is in unhealthy state. Please check Ceph cluster health or RGW connection.',
+              description: 'Cluster Object Store is in unhealthy state for more than %s. Please check Ceph cluster health or RGW connection.' % $._config.clusterObjectStoreStateAlertTime,
```
You can add endpoint details and state from the labels in metrics. That'll be more helpful.

anmolsachan replied:
Makes sense

umangachapagain commented on metrics/mixin/alerts/services.libsonnet:
```diff
@@ -0,0 +1,27 @@
+            annotations: {
+              message: 'Cluster Object Store is in unhealthy state. Please check Ceph cluster health.',
+              description: 'Cluster Object Store is in unhealthy state for more than %s. Please check Ceph cluster health.' % $._config.clusterObjectStoreStateAlertTime,
+              storage_type: $._config.objectStorageType,
+              severity_level: 'error',
```
remove these. we don't use it

anmolsachan replied:
We do, Its required for UI

umangachapagain commented on metrics/mixin/alerts/services.libsonnet:
```diff
@@ -0,0 +1,27 @@
+            },
+            annotations: {
+              message: 'Cluster Object Store is in unhealthy state. Please check Ceph cluster health.',
+              description: 'Cluster Object Store is in unhealthy state for more than %s. Please check Ceph cluster health.' % $._config.clusterObjectStoreStateAlertTime,
```
add labels from metrics for more context.

anmolsachan replied:
ack

umangachapagain commented on metrics/mixin/alerts/services-external.libsonnet:
```diff
@@ -0,0 +1,27 @@
+            annotations: {
+              message: 'Cluster Object Store is in unhealthy state. Please check Ceph cluster health or RGW connection.',
+              description: 'Cluster Object Store is in unhealthy state for more than %s. Please check Ceph cluster health or RGW connection.' % $._config.clusterObjectStoreStateAlertTime,
+              storage_type: $._config.objectStorageType,
+              severity_level: 'error',
```
We can get rid of this.

anmolsachan replied:
Yeah we can, but let's keep all the alerts related to OCS same. These will differ from the rook ones then, might create confusion.

umangachapagain commented on metrics/mixin/build/example.jsonnet:
```diff
@@ -0,0 +1,7 @@
+local kp = (import 'jsonnet/kube-prometheus.libsonnet') + {
```
can we rename file to `internal.jsonnet`?

anmolsachan replied:
sure

umangachapagain commented on metrics/mixin/build/jsonnet/kube-prometheus.libsonnet:
```diff
@@ -0,0 +1,18 @@
+local k = import '../vendor/ksonnet/ksonnet.beta.3/k.libsonnet';
+local configMapList = k.core.v1.configMapList;
```
Do we need this??

anmolsachan replied:
No, will remove

umangachapagain commented on metrics/mixin/rules/rules.libsonnet:
```diff
@@ -0,0 +1,11 @@
+    groups+: [
+      {
+        name: 'ocs.rules',
+        rules: [
```
Let's not keep it empty. 
Remove it and add when required.

anmolsachan replied:
Cool

umangachapagain commented:
What is `metrics/mixin/prometheus_rules.yaml` file? It shows no changes. Can we get rid of it?

anmolsachan commented on metrics/mixin/build/example-external.jsonnet:
```diff
@@ -0,0 +1,7 @@
+local kp = (import 'jsonnet/kube-prometheus-external.libsonnet') + {
```
sure

anmolsachan commented on metrics/mixin/build/jsonnet/kube-prometheus-external.libsonnet:
```diff
@@ -0,0 +1,18 @@
+local k = import '../vendor/ksonnet/ksonnet.beta.3/k.libsonnet';
+local configMapList = k.core.v1.configMapList;
```
No, will remove

anmolsachan commented:
> What is `metrics/mixin/prometheus_rules.yaml` file? It shows no changes. Can we get rid of it?

Done !.

anmolsachan commented:
/retest

jarrpa commented:
With #739 merged, please update this one and we'll merge it straightaway.

anmolsachan commented:
/retest

anmolsachan commented:
@jarrpa The PR is rebased. Thanks.
